// @cnpm/utils/src/account.js
var cookieUtil = require('@cnpm/cookie-util');
var Utils = require('./util');

/** 登录网址 */
var URL_LOGIN = 'https://passport.chinaso.com/login';

/** 退出地址 */
var URL_EXIT = 'https://passport.chinaso.com/logout';

/** 登录跳转地址 */
var LOGIN_URL = URL_LOGIN + '?service=' + encodeURIComponent(window.location.href);

/** 用户中心 */
var CONF_URL = 'http://my.chinaso.com/';


function Account(setting) {
  this.options = Utils.extend({
    /** 登录页面的地址 */
    loginURL: LOGIN_URL,
    /** 用户退出时发送的异步请求地址 */
    exitURL: URL_EXIT,
    /** 用户中心 */
    confURL: CONF_URL,
    /** 登录账号 cookie 名 */
    cookieLogin: 'chinasoticket_account',
    /** 登录用户名 */
    cookieUserName: 'chinasoticket_titlename'
  }, setting);
}

Utils.extend(Account.prototype, {

  /**
   * 判断是否为主页
   * 利用主页面特有的 meta 标记为依据
   * <meta name="iswwwhomepage" content="yes">
   * @returns {Boolean} true 为是主页，否则不是主页
   */
  isHomePage: function isHomePage() {
    return $('meta[name=iswwwhomepage]').length > 0;
  },

  /**
   * 判断是否登录
   * @returns {Boolean} true 为登录，否则为未登录
   */
  isLogin: function isLogin() {
    return !!cookieUtil.get(this.options.cookieLogin, true);
  },

  /**
   * 账号退出
   * 学术频道的退出操作比较特殊，在这里判断特殊处理
   */
  exit: function exit() {
    var T = this;
    // 学术频道特殊退出方式
    if (window.location.href.indexOf('scholar.chinaso.com') > 0) {
      location.href = T.options.exitURL + '?service=' + encodeURIComponent('http://scholar.chinaso.com/api/index.php?m=home&c=UserLogout&a=logOutDirect&sourceUrl=' + window.location.href);
    } else {
      location.href = T.options.exitURL + '?service=' + encodeURIComponent(window.location.href);
    }
  }
});

module.exports = Account;
